storage/testonly/fake_node_reader.go

```diff
@@ -0,0 +1,196 @@
+package testonly
+
+import (
+	"fmt"
+
+	"encoding/hex"
+	"github.com/google/trillian/crypto"
+	"github.com/google/trillian/merkle"
+	"github.com/google/trillian/storage"
+)
+
+// This is a fake implementation of a NodeReader intended for use in testing Merkle path code.
+// Building node sets for tests by hand is onerous and error prone, especially when trying
+// to test code reading from multiple tree revisions. It cannot live in the main testonly
+// package as this creates import cycles.
+
+// NodeMapping is a struct we use because we can't use NodeIDs as map keys. Callers pass this
+// and FakeNodeReader internally manages derived keys.
+type NodeMapping struct {
+	NodeID storage.NodeID
+	Node   storage.Node
+}
+
+// FakeNodeReader is an implementation of storage.NodeReader that's preloaded with a set of
+// NodeID -> Node mappings and will return only those. Requesting any other nodes results in
+// an error. For use in tests only, does not implement any other storage APIs.
+type FakeNodeReader struct {
+	treeSize     int64
+	treeRevision int64
+	nodeMap      map[string]storage.Node
+}
+
+// NewFakeNodeReader creates and returns a FakeNodeReader with the supplied nodeID -> Node
+// mappings assuming that all the nodes are at a specified tree revision. All the nodeIDs
+// must be distinct.
+func NewFakeNodeReader(mappings []NodeMapping, treeSize, treeRevision int64) *FakeNodeReader {
+	nodeMap := make(map[string]storage.Node)
+
+	for _, mapping := range mappings {
+		_, ok := nodeMap[mapping.NodeID.String()]
+
+		if ok {
+			// Duplicate mapping - bail out
+			return nil
+		}
+
+		nodeMap[mapping.NodeID.String()] = mapping.Node
+	}
+
+	return &FakeNodeReader{nodeMap: nodeMap, treeSize: treeSize, treeRevision: treeRevision}
+}
+
+// GetTreeRevisionAtSize implements the corresponding NodeReader API
+func (f FakeNodeReader) GetTreeRevisionAtSize(treeSize int64) (int64, error) {
+	if f.treeSize != treeSize {
+		return int64(0), fmt.Errorf("GetTreeRevisionAtSize() got treeSize:%d, want: %d", treeSize, f.treeSize)
+	}
+
+	return f.treeRevision, nil
+}
+
+// GetMerkleNodes implements the corresponding NodeReader API
+func (f FakeNodeReader) GetMerkleNodes(treeRevision int64, NodeIDs []storage.NodeID) ([]storage.Node, error) {
+	if f.treeRevision > treeRevision {
+		return []storage.Node{}, fmt.Errorf("GetMerkleNodes() got treeRevision:%d, want up to: %d", treeRevision, f.treeRevision)
```

Comment:
Suggestion: Return nil instead of []storage.Node{}? Less characters, type-change proof.

(Same for other occurrences, if you take if, ofc.)